File: platforms/flutter/android/unitrack_sdk/src/main/java/com/unitrack/sdk/providers/HttpProvider.kt
package com.unitrack.sdk.providers

import android.app.Application
import android.util.Log
import com.unitrack.sdk.UniTrack
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicLong

/**
 * Generic HTTP analytics provider.
 *
 * Mục tiêu: gắn UniTrack lên Kibana / Elasticsearch / Logstash / OpenSearch /
 * backend FPT nội bộ trong 5 dòng config — không cần code transport, retry,
 * batch, offline. UniTrack lo hết qua [PendingQueue].
 *
 *   UniTrack.addHttpProvider(
 *       id        = "kibana",
 *       endpoint  = "https://kibana.fpt.vn/_bulk",
 *       format    = PayloadFormat.ELASTIC_BULK,
 *       headers   = mapOf("Authorization" to "ApiKey ..."),
 *       batchSize = 50,
 *   )
 *
 * Behaviour:
 *   - 2xx                   → SUCCESS (queue clears event for this provider)
 *   - 4xx (except 408/429)  → DROP    (schema sai — đừng loop)
 *   - 5xx / 408 / 429       → RETRY   (PendingQueue backoff exponential)
 *   - network/timeout       → RETRY
 *
 * Stamp session_id từ UniTrack core vào mọi event tự động.
 */
class HttpProvider(
    private val id: String,
    private val endpoint: String,
    private val format: PayloadFormat = PayloadFormat.JSON_SINGLE,
    private val headers: Map<String, String> = emptyMap(),
    private val batchSize: Int = 50,
    private val flushIntervalMs: Long = 30_000,
    private val connectTimeoutMs: Int = 10_000,
    private val readTimeoutMs: Int = 15_000,
) : AnalyticsProvider {

    override val providerId: String get() = id

    private val pending = ConcurrentLinkedQueue<JSONObject>()
    private val lastFlushAt = AtomicLong(0)
    private val exec = Executors.newSingleThreadExecutor { r ->
        Thread(r, "ut-http-$id").apply { isDaemon = true }
    }
    private val resolvedHeaders: Map<String, String> = resolveHeaderSecrets(headers, id)

    override fun initialize(app: Application) { /* no SDK to wake */ }
    override fun track(name: String, properties: Map<String, Any?>) {
        // Default fall-through (send() is the real entry point).
        send(name, properties)
    }
    override fun setUser(userId: String?, traits: Map<String, Any?>) { /* no-op */ }
    override fun setScreen(name: String) { /* no-op */ }

    override fun send(name: String, properties: Map<String, Any?>): ProviderResult {
        val event = stamp(name, properties)
        return when (format) {
            PayloadFormat.JSON_SINGLE -> postOne(event)
            PayloadFormat.JSON_LINES,
            PayloadFormat.JSON_ARRAY,
            PayloadFormat.ELASTIC_BULK -> bufferAndMaybeFlush(event)
        }
    }

    private fun stamp(name: String, properties: Map<String, Any?>): JSONObject {
        val o = JSONObject(properties.mapValues { it.value ?: JSONObject.NULL })
        o.put("event_name", name)
        try {
            o.put("session_id", UniTrack.currentSessionId())
            o.put("session_index", UniTrack.sessionIndex())
        } catch (_: Throwable) { /* before init */ }
        if (!o.has("timestamp")) o.put("timestamp", System.currentTimeMillis())
        // W3C: stamp trace_id ở event-level khi tracing bật ở remote config,
        // join với backend log / Snowplow / Firebase qua cùng 1 mã. HTTP-header
        // injection vẫn do UniTrackTracingInterceptor lo allowlist riêng.
        try {
            if (UniTrack.tracingSnapshot().enabled && !o.has("trace_id")) {
                val (traceId, spanId) =
                    com.unitrack.sdk.bridge.NativeBridge.newTrace()
                o.put("trace_id", traceId)
                o.put("span_id",  spanId)
            }
        } catch (_: Throwable) { /* native not loaded yet */ }
        return o
    }

    private fun bufferAndMaybeFlush(event: JSONObject): ProviderResult {
        pending.add(event)
        val now = System.currentTimeMillis()
        val due = pending.size >= batchSize ||
                  (now - lastFlushAt.get()) >= flushIntervalMs
        if (!due) return ProviderResult.SUCCESS  // batched, will leave on next flush
        return flushBatch()
    }

    /** Force a flush — exposed for tests / lifecycle. */
    fun flush(): ProviderResult = flushBatch()

    private fun flushBatch(): ProviderResult {
        val batch = mutableListOf<JSONObject>()
        while (batch.size < batchSize) {
            batch.add(pending.poll() ?: break)
        }
        if (batch.isEmpty()) return ProviderResult.SUCCESS
        lastFlushAt.set(System.currentTimeMillis())
        val body = when (format) {
            PayloadFormat.JSON_LINES   -> batch.joinToString("\n") { it.toString() }
            PayloadFormat.JSON_ARRAY   -> JSONArray(batch).toString()
            PayloadFormat.ELASTIC_BULK -> batch.joinToString("\n") {
                """{"index":{}}""" + "\n" + it.toString()
            } + "\n"
            PayloadFormat.JSON_SINGLE  -> batch.first().toString()
        }
        val contentType = when (format) {
            PayloadFormat.ELASTIC_BULK, PayloadFormat.JSON_LINES -> "application/x-ndjson"
            else -> "application/json"
        }
        val r = post(body, contentType)
        if (r != ProviderResult.SUCCESS) {
            // Put events back so PendingQueue replays them; we don't want to
            // double-buffer between in-memory pending + on-disk PendingQueue.
            batch.forEach { pending.offer(it) }
        }
        return r
    }

    private fun postOne(event: JSONObject): ProviderResult {
        return post(event.toString(), "application/json")
    }

    private fun post(body: String, contentType: String): ProviderResult {
        var conn: HttpURLConnection? = null
        return try {
            conn = (URL(endpoint).openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                connectTimeout = connectTimeoutMs
                readTimeout    = readTimeoutMs
                doOutput = true
                setRequestProperty("Content-Type", contentType)
                for ((k, v) in resolvedHeaders) setRequestProperty(k, v)
            }
            conn.outputStream.use { it.write(body.toByteArray(Charsets.UTF_8)) }
            val code = conn.responseCode
            when {
                code in 200..299           -> ProviderResult.SUCCESS
                code == 408 || code == 429 -> ProviderResult.RETRY
                code in 500..599           -> ProviderResult.RETRY
                code in 400..499           -> {
                    Log.w("UTHttpProvider", "$id 4xx ($code) → DROP")
                    ProviderResult.DROP
                }
                else                       -> ProviderResult.RETRY
            }
        } catch (e: Throwable) {
            Log.w("UTHttpProvider", "$id network error: ${e.message} → RETRY")
            ProviderResult.RETRY
        } finally {
            conn?.disconnect()
        }
    }

    companion object {
        /**
         * Resolve `${ENV_FOO}` placeholders trong header VALUES bằng env vars.
         * Cho phép Portal lưu reference non-secret kiểu `${ENV_KIBANA_KEY}`
         * thay vì raw API key — app/process set thật ở launch (gradle BuildConfig
         * → System.setProperty, hoặc k8s/Docker env). Unmatched → giữ literal,
         * log 1 lần. Header KEYS không đổi.
         */
        private val PLACEHOLDER = Regex("""\$\{ENV_([A-Z0-9_]+)\}""")

        @JvmStatic
        fun resolveHeaderSecrets(headers: Map<String, String>, providerId: String): Map<String, String> {
            val env = System.getenv()
            return headers.mapValues { (k, v) ->
                PLACEHOLDER.replace(v) { m ->
                    val name = "ENV_" + m.groupValues[1]
                    env[name] ?: run {
                        Log.w("UTHttpProvider", "$providerId header $k missing env $name")
                        m.value
                    }
                }
            }
        }
    }
}

enum class PayloadFormat {
    /** 1 event = 1 POST as a single JSON object. Simplest backends. */
    JSON_SINGLE,
    /** Batched: NDJSON, 1 line per event. */
    JSON_LINES,
    /** Batched: a JSON array per POST. */
    JSON_ARRAY,
    /** Elasticsearch `_bulk` API: action line + doc line, NDJSON, trailing newline. */
    ELASTIC_BULK,
}

File: platforms/flutter/ios/Native/swift/Providers/HttpProvider.swift
// HttpProvider — generic HTTP analytics provider.
//
// Mục tiêu: gắn UniTrack lên Kibana / Elasticsearch / Logstash / OpenSearch /
// backend FPT nội bộ trong 5 dòng config — không cần code transport, retry,
// batch, offline. UniTrack lo hết qua `PendingQueue`.
//
//   UniTrack.addHttpProvider(
//       id: "kibana",
//       endpoint: URL(string: "https://kibana.fpt.vn/_bulk")!,
//       format: .elasticBulk,
//       headers: ["Authorization": "ApiKey ..."],
//       batchSize: 50)
//
// Behaviour:
//   - 2xx                   → .success
//   - 4xx (except 408/429)  → .drop    (schema sai — đừng loop)
//   - 5xx / 408 / 429       → .retry   (PendingQueue backoff exponential)
//   - network/timeout       → .retry
import Foundation

public enum PayloadFormat: Int {
    /// 1 event = 1 POST as a single JSON object. Simplest backends.
    case jsonSingle = 0
    /// Batched: NDJSON, 1 line per event.
    case jsonLines = 1
    /// Batched: a JSON array per POST.
    case jsonArray = 2
    /// Elasticsearch _bulk API: action line + doc line, NDJSON, trailing newline.
    case elasticBulk = 3
}

public final class HttpProvider: AnalyticsProvider {

    public let providerId: String

    private let endpoint: URL
    private let format: PayloadFormat
    private let headers: [String: String]
    private let batchSize: Int
    private let flushInterval: TimeInterval

    private let q = DispatchQueue(label: "ut.httpprovider", qos: .utility)
    private var pending: [[String: Any]] = []
    private var lastFlushAt: TimeInterval = 0
    private let session: URLSession
    private let resolvedHeaders: [String: String]

    public init(id: String,
                endpoint: URL,
                format: PayloadFormat = .jsonSingle,
                headers: [String: String] = [:],
                batchSize: Int = 50,
                flushInterval: TimeInterval = 30,
                connectTimeout: TimeInterval = 10,
                readTimeout: TimeInterval = 15) {
        self.providerId    = id
        self.endpoint      = endpoint
        self.format        = format
        self.headers       = headers
        self.resolvedHeaders = Self.resolveHeaderSecrets(headers, providerId: id)
        self.batchSize     = batchSize
        self.flushInterval = flushInterval
        let cfg = URLSessionConfiguration.default
        cfg.timeoutIntervalForRequest  = readTimeout
        cfg.timeoutIntervalForResource = readTimeout + connectTimeout
        // Skip UniTrackURLProtocol so we don't recursively track our own POSTs.
        cfg.protocolClasses = (cfg.protocolClasses ?? []).filter {
            String(describing: $0) != "UniTrackURLProtocol"
        }
        self.session = URLSession(configuration: cfg)
    }

    public func initializeProvider() { /* no SDK to wake */ }

    public func track(_ name: String, _ properties: [String: Any]) {
        _ = send(name, properties)
    }

    public func setUser(_ userId: String?, _ traits: [String: Any]) { /* no-op */ }
    public func setScreen(_ name: String) { /* no-op */ }

    public func send(_ name: String, _ properties: [String: Any]) -> ProviderResult {
        let event = stamp(name: name, properties: properties)
        switch format {
        case .jsonSingle:
            return postSync(body: jsonData(event), contentType: "application/json")
        case .jsonLines, .jsonArray, .elasticBulk:
            return bufferAndMaybeFlush(event: event)
        }
    }

    /// Force a flush — exposed for tests / lifecycle.
    public func flush() -> ProviderResult { flushBatch() }

    private func stamp(name: String, properties: [String: Any]) -> [String: Any] {
        var o = properties
        o["event_name"]    = name
        o["session_id"]    = UniTrack.currentSessionId()
        o["session_index"] = UniTrack.sessionIndex()
        if o["timestamp"] == nil {
            o["timestamp"] = Int(Date().timeIntervalSince1970 * 1000)
        }
        // W3C: stamp a fresh root trace per event so the kho event row joins
        // with backend log / Snowplow / Firebase using the same trace_id. Only
        // when tracing is enabled in remote config — avoids minting noise when
        // the operator hasn't opted into W3C yet. Event-level only; HTTP-header
        // injection is governed separately by UniTrackURLProtocol's allowlist.
        if UniTrackRemoteConfig.latest?.tracing?.enabled == true,
           o["trace_id"] == nil {
            let ids = UniTrackTracing.newTrace()
            o["trace_id"] = ids.traceId
            o["span_id"]  = ids.spanId
        }
        return o
    }

    /// Resolve `${ENV_FOO}` placeholders in header VALUES against ProcessInfo's
    /// environment. Lets the Portal store a non-secret reference like
    /// `${ENV_KIBANA_KEY}` instead of the raw API key — app/process sets the
    /// real value at launch (xcconfig / scheme env / k8s secret).
    ///
    /// Unmatched placeholder → keep literal, log once. Multiple placeholders
    /// in one value all resolve. Header KEYS stay unchanged.
    static func resolveHeaderSecrets(_ headers: [String: String],
                                     providerId: String) -> [String: String] {
        let env = ProcessInfo.processInfo.environment
        var out: [String: String] = [:]
        let pattern = try? NSRegularExpression(pattern: #"\$\{ENV_([A-Z0-9_]+)\}"#)
        for (k, v) in headers {
            guard let re = pattern else { out[k] = v; continue }
            var resolved = v
            let range = NSRange(v.startIndex..<v.endIndex, in: v)
            let matches = re.matches(in: v, range: range).reversed()
            for m in matches {
                guard let full = Range(m.range, in: v),
                      let nameR = Range(m.range(at: 1), in: v) else { continue }
                let name = "ENV_" + String(v[nameR])
                if let val = env[name] {
                    resolved.replaceSubrange(full, with: val)
                } else {
                    NSLog("[UniTrack] HttpProvider %@ header %@ missing env %@",
                          providerId, k, name)
                }
            }
            out[k] = resolved
        }
        return out
    }

    private func bufferAndMaybeFlush(event: [String: Any]) -> ProviderResult {
        var due = false
        q.sync {
            pending.append(event)
            let now = Date().timeIntervalSince1970
            if pending.count >= batchSize || (now - lastFlushAt) >= flushInterval {
                due = true
            }
        }
        return due ? flushBatch() : .success
    }

    private func flushBatch() -> ProviderResult {
        var batch: [[String: Any]] = []
        q.sync {
            let n = min(batchSize, pending.count)
            batch = Array(pending.prefix(n))
            pending.removeFirst(n)
            lastFlushAt = Date().timeIntervalSince1970
        }
        guard !batch.isEmpty else { return .success }

        let (body, contentType) = encode(batch: batch)
        let r = postSync(body: body, contentType: contentType)
        if r != .success {
            // Put events back so PendingQueue replays them; avoid double-buffering.
            q.sync { pending.insert(contentsOf: batch, at: 0) }
        }
        return r
    }

    private func encode(batch: [[String: Any]]) -> (Data, String) {
        switch format {
        case .jsonSingle:
            return (jsonData(batch.first ?? [:]), "application/json")
        case .jsonLines:
            let s = batch.map { String(data: jsonData($0), encoding: .utf8) ?? "{}" }
                .joined(separator: "\n") + "\n"
            return (Data(s.utf8), "application/x-ndjson")
        case .jsonArray:
            let arr = try? JSONSerialization.data(withJSONObject: batch)
            return (arr ?? Data("[]".utf8), "application/json")
        case .elasticBulk:
            var s = ""
            for e in batch {
                s += "{\"index\":{}}\n"
                s += (String(data: jsonData(e), encoding: .utf8) ?? "{}") + "\n"
            }
            return (Data(s.utf8), "application/x-ndjson")
        }
    }

    private func jsonData(_ o: [String: Any]) -> Data {
        (try? JSONSerialization.data(withJSONObject: o)) ?? Data("{}".utf8)
    }

    private func postSync(body: Data, contentType: String) -> ProviderResult {
        var req = URLRequest(url: endpoint)
        req.httpMethod = "POST"
        req.setValue(contentType, forHTTPHeaderField: "Content-Type")
        for (k, v) in resolvedHeaders { req.setValue(v, forHTTPHeaderField: k) }
        req.httpBody = body

        let sem = DispatchSemaphore(value: 0)
        var result: ProviderResult = .retry
        let task = session.dataTask(with: req) { _, resp, err in
            defer { sem.signal() }
            if let e = err {
                NSLog("[UniTrack] HttpProvider %@ network error: %@ → RETRY",
                      self.providerId, e.localizedDescription)
                result = .retry
                return
            }
            let code = (resp as? HTTPURLResponse)?.statusCode ?? 0
            switch code {
            case 200...299:                   result = .success
            case 408, 429, 500...599:         result = .retry
            case 400...499:
                NSLog("[UniTrack] HttpProvider %@ 4xx (%d) → DROP",
                      self.providerId, code)
                result = .drop
            default:                          result = .retry
            }
        }
        task.resume()
        _ = sem.wait(timeout: .now() + 30)
        return result
    }
}
